nomt/src/cursor.rs

```diff
@@ -95,95 +110,147 @@ impl PageCacheCursor {
-            .set(self.depth as usize, true);
+            .set(self.depth as usize, bit);
         self.depth += 1;
     }
 
-    /// Returns the current location of the cursor, represented as a tuple of the key path and the
-    /// depth.
-    pub fn location(&self) -> (KeyPath, u8) {
-        (self.path, self.depth)
+    /// Traverse to the sibling node of the current position. No-op at the root.
+    pub fn sibling(&mut self) {
+        if self.depth == 0 {
+            return;
+        }
+
+        let bits = self.path.view_bits_mut::<Msb0>();
+        let i = self.depth as usize - 1;
+        bits.set(i, !bits[i]);
     }
 
     /// Returns the node at the current location of the cursor.
-    pub fn node(&self, page_cache: &PageCache) -> Node {
-        if self.depth == 0 {
-            self.root
-        } else {
-            // Calculate the page ID of the current page.
-            let cur_page_id = PageIdsIterator::new(&self.path)
-                .last()
-                .expect("the cursor is not at the root");
-            // Check if the page is already cached locally by checking it's page ID. If it's not,
-            // retrieve the page from the cache and update the cache.
-            let mut cached_page_guard = self.cached_page.borrow_mut();
-            let page = match &*cached_page_guard {
-                Some((cached_page_id, cached_page)) => {
-                    if *cached_page_id == cur_page_id {
-                        cached_page.clone()
-                    } else {
-                        let page = page_cache.retrieve(cur_page_id).wait();
-                        *cached_page_guard = Some((cur_page_id, page.clone()));
-                        page
-                    }
-                }
-                _ => {
-                    let page = page_cache.retrieve(cur_page_id).wait();
-                    *cached_page_guard = Some((cur_page_id, page.clone()));
-                    page
-                }
-            };
-            // Now having the page, we can extract the node from it.
-            let page_path = last_page_path(&self.path, self.depth);
-            let node_index = node_index(page_path);
-            page.node(node_index)
+    pub fn node(&self) -> Node {
+        match self.cached_page {
+            None => self.root,
+            Some((_, ref page)) => {
+                let path = last_page_path(&self.path, self.depth);
+                page.node(node_index(path))
+            }
+        }
+    }
+
+    /// Peek at the sibling node of the current position without moving the cursor. At the root,
+    /// gives the terminator.
+    pub fn peek_sibling(&self) -> Node {
+        match self.cached_page {
+            None => trie::TERMINATOR,
+            Some((_, ref page)) => {
+                let path = last_page_path(&self.path, self.depth);
+                page.node(sibling_index(path))
+            }
         }
     }
 }
 
+impl CursorApi for PageCacheCursor {
```

Comment:
A trick to avoid renamings at low cost is to use the `nomt_core::Cursor`.

Reply:
that is, rather than importing `Cursor as CursorApi` you suggest `impl nomt_core::Cursor`?

I will try it, it does look better. non-blocking

Reply:
https://github.com/thrumdev/nomt/pull/51